File: design_analysis_algorithms/question C user profile hash table/user_profile_hash _table.py
class UserHashTable:
    def __init__(self, size=100):
        # Initialize hash table with a given size
        self.size = size
        self.table = [[] for _ in range(size)]  # Create a list of empty lists for separate chaining

    def _hash(self, username: str) -> int:
        # Simple hash function: sum of ASCII values of the characters mod table size
        return sum(ord(c) for c in username) % self.size

    def add_user(self, username: str, profile_data: dict):
        # Hash the username to find the index in the table
        index = self._hash(username)
        
        # Check if the username already exists; if so, update the profile
        for entry in self.table[index]:
            if entry[0] == username:
                entry[1] = profile_data
                return
        
        # If the username doesn't exist, add a new entry
        self.table[index].append((username, profile_data))

    def get_user(self, username: str) -> dict:
        # Hash the username to find the index
        index = self._hash(username)
        
        # Look for the username in the corresponding chain (linked list)
        for entry in self.table[index]:
            if entry[0] == username:
                return entry[1]  # Return the profile data
        
        # If not found, return an empty dictionary (or could raise an exception)
        return {}

# Example Usage
user_table = UserHashTable()
user_table.add_user("alice", {"age": 25, "location": "NYC"})
user_table.add_user("bob", {"age": 30, "location": "LA"})

print(user_table.get_user("alice"))  # Output: {'age': 25, 'location': 'NYC'}
print(user_table.get_user("bob"))  # Output: {'age': 30, 'location': 'LA'}
print(user_table.get_user("charlie"))  # Output: {}

File: design_analysis_algorithms/question C user profile hash table/user_profile_ hash_ table.cpp
#include <iostream>
#include <vector>
#include <list>
#include <string>
#include <unordered_map>

using namespace std;

class UserHashTable {
private:
    // Hash table using separate chaining with linked lists
    vector<list<pair<string, unordered_map<string, string>>>> table;
    int size;

    // Hash function to convert username to an index
    int hash(const string& username) {
        int hashValue = 0;
        for (char c : username) {
            hashValue += c;
        }
        return hashValue % size;
    }

public:
    UserHashTable(int size = 100) : size(size) {
        table.resize(size);  // Initialize the table with empty lists
    }

    // Add or update user profile
    void addUser(const string& username, const unordered_map<string, string>& profileData) {
        int index = hash(username);

        // Search for the username in the list at the calculated index
        for (auto& entry : table[index]) {
            if (entry.first == username) {
                // Username found, update the profile data
                entry.second = profileData;
                return;
            }
        }

        // If username not found, add a new entry
        table[index].push_back(make_pair(username, profileData));
    }

    // Get user profile data by username
    unordered_map<string, string> getUser(const string& username) {
        int index = hash(username);

        // Search for the username in the list at the calculated index
        for (auto& entry : table[index]) {
            if (entry.first == username) {
                return entry.second;  // Return the profile data
            }
        }

        // If not found, return an empty map
        return unordered_map<string, string>();
    }
};

// Example Usage
int main() {
    UserHashTable userTable;

    unordered_map<string, string> aliceProfile = {{"age", "25"}, {"location", "NYC"}};
    unordered_map<string, string> bobProfile = {{"age", "30"}, {"location", "LA"}};

    userTable.addUser("alice", aliceProfile);
    userTable.addUser("bob", bobProfile);

    unordered_map<string, string> alice = userTable.getUser("alice");
    unordered_map<string, string> bob = userTable.getUser("bob");

    cout << "Alice's profile: ";
    for (const auto& pair : alice) {
        cout << pair.first << ": " << pair.second << ", ";
    }
    cout << endl;

    cout << "Bob's profile: ";
    for (const auto& pair : bob) {
        cout << pair.first << ": " << pair.second << ", ";
    }
    cout << endl;

    return 0;
}
